### PR TITLE
Feature document model

### PR DESCRIPTION
**Pull request by gaimanjing**

### CODE REVIEW

File: src/Model/DocumentModel.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+{
+public:
+  DocumentModel(const nlohmann::json& schema, const nlohmann::json& document);
+  ~DocumentModel();
+
+  void addAt(const json::json_pointer& path, const json& value);
+  void replaceAt(const json::json_pointer& path, const json& value);
+  void deleteAt(const json::json_pointer& path);
+
+  json documentJson();
```
Comment (harry75369) [inline src/Model/DocumentModel.hpp:17]: 1. should add "const" suffix to declaration if it is a copy
2. however, copy of the whole document is less efficient

Consider redesign this API for retrieving data.

File: src/Model/JsonSchemaValidator.hpp [hunk 2]
```diff
@@ -4,18 +4,28 @@
...
 
   void setRootSchema(const nlohmann::json& schemaJson);
-  bool validate(const nlohmann::json& targetJson);
+  bool validate(const nlohmann::json& targetDocument);
+  bool validate(const std::string& className, const nlohmann::json& targetDocument);
 
 private:
   valijson::Validator validator_;
   valijson::Schema schema_;
+  std::map<std::string, std::string> classTitleMap_;
```
Comment (harry75369) [inline src/Model/JsonSchemaValidator.hpp:22]: Consider using unordered_map for better performance

**Review by harry75369 — CHANGES_REQUESTED** [review]

File: src/Model/JsonSchemaValidator.cpp [hunk 3]
```diff
@@ -15,24 +16,33 @@ JsonSchemaValidator::~JsonSchemaValidator()
...
+
+bool JsonSchemaValidator::validate(const valijson::Subschema* subschema,
+                                   const nlohmann::json& targetDocument)
+{
+  if (!subschema)
+  {
+    return false;
+  }
+
   ValidationResults results;
```
Comment (harry75369) [inline src/Model/JsonSchemaValidator.cpp:43]: Better if this type has `valijson` namespace.

File: src/Model/DocumentModel.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+DocumentModel::DocumentModel(const nlohmann::json& schema, const nlohmann::json& document)
+{
+  m_validator.setRootSchema(schema);
+  if (m_validator.validate(document))
+  {
+    from_json(document, m_doc);
+  }
+  else
+  {
+    throw std::runtime_error("Invalid document");
```
Comment (harry75369) [inline src/Model/DocumentModel.cpp:16]: It is not a good practice to throw exceptions in constructor.

File: src/Model/DocumentModel.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  {
+    return;
+  }
+  auto& current = currentPath.back();
+  calculateRelativePath(ancestorPath, currentPath.parent_pointer(), relativePath);
+  relativePath.push_back(current);
+}
+
+void DocumentModel::save()
+{
```
Comment (harry75369) [inline src/Model/DocumentModel.cpp:146]: empty?

File: src/Model/DesignDocumentModel.hpp [hunk 6]
```diff
@@ -0,0 +1,2450 @@
...
+ *
+ * Defines an abstract rectangle.
+ *
+ * Frame information for the object. When the object is a group, its frame is the union of
+ * all objects in the group, whether or not the group contains a mask. This property stores
+ * information about the object after the matrix transformation.
+ */
+struct Rect
+{
+  nlohmann::json rect_class;
```
Comment (harry75369) [inline src/Model/DesignDocumentModel.hpp:149]: Could we handle such const field properly?

File: src/Model/DocumentModel.cpp (hunk 5 above)
Comment (gaimanjing) [reply src/Model/DocumentModel.cpp:146]: It is a work in progress, saving vgg work to backend.

File: src/Model/DesignDocumentModel.hpp (hunk 6 above)
Comment (gaimanjing) [reply src/Model/DesignDocumentModel.hpp:149]: Yes, we will handle it carefully. 
Construct this struct from json need public field, so const filed is not possible. 
The field's value will be validate if changed.

File: src/Model/JsonSchemaValidator.cpp [hunk 7]
```diff
@@ -15,24 +16,33 @@ JsonSchemaValidator::~JsonSchemaValidator()
...
-  try
-  {
-    parser.populateSchema(schemaDocumentAdapter, schema_);
-  }
-  catch (std::exception& e)
-  {
-    WARN("#vgg json schema set schema error: %s", e.what());
-    throw;
-  }
+  auto tmp_json = schemaJson;
```
[the remark this answers is not in the document]
Comment (gaimanjing) [reply src/Model/JsonSchemaValidator.cpp:19]: Fixed, remove const.

File: src/Model/DocumentModel.cpp (hunk 4 above)
Comment (gaimanjing) [reply src/Model/DocumentModel.cpp:16]: Fixed, remove validating, move up validator setup.

**Pull request merged**